File: set-matrix-zeroes.cpp
#include <vector>

using namespace std;

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
		int rows = matrix.size();
		int cols = matrix[0].size();

		bool rflag = false;
		bool cflag = false;

		for (int r=0; r<rows; ++r) {
			if (matrix[r][0] == 0) {
				cflag = true;
				break;
			}
		}

		for (int c=0; c<cols; ++c) {
			if (matrix[0][c] == 0) {
				rflag = true;
				break;
			}
		}

		for (int r=1; r<rows; ++r) {
			for (int c=1; c<cols; ++c) {
				if (matrix[r][c] == 0) {
					matrix[r][0] = 0;
					matrix[0][c] = 0;
				}
			}
		}

		for (int r=1; r<rows; ++r) {
			for (int c=1; c<cols; ++c) {
				if ((matrix[r][0] == 0) || (matrix[0][c] ==0)) {
					matrix[r][c] = 0;
				}
			}
		}

		if (rflag) {
			for (int c=0; c<cols; ++c) {
				matrix[0][c] = 0;
			}
		}

		if (cflag) {
			for (int r=0; r<rows; ++r) {
				matrix[r][0] = 0;
			}
		}
    }
};

File: set-matrix-zeroes.py
from typing import List


class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        rows, cols = len(matrix), len(matrix[0])
        rflag, cflag = False, False

        for r in range(rows):
            if matrix[r][0] == 0:
                cflag = True
                break
        
        for c in range(cols):
            if matrix[0][c] == 0:
                rflag = True
                break

        for r in range(1, rows):
            for c in range(1, cols):
                if matrix[r][c] == 0:
                    matrix[0][c] = 0
                    matrix[r][0] = 0

        for r in range(1, rows):
            for c in range(1, cols):
                if matrix[r][0] == 0 or matrix[0][c] == 0:
                    matrix[r][c] = 0

        if rflag:
            for c in range(cols):
                matrix[0][c] = 0

        if cflag:
            for r in range(rows):
                matrix[r][0] = 0
